File: main_app/ios/Curtaincall/CurtaincallWidget/WidgetDataLoader.swift
import Foundation

struct WidgetCard {
    let cardId: Int
    let quote: String
    let workTitle: String
}

enum WidgetDataLoader {
    // Duplicated from Curtaincall/Config.swift — the widget target intentionally
    // doesn't import the app's networking layer (keeps it independently buildable).
    private static let supabaseURL = URL(string: "https://hixymiidpxnnovtmsvfp.supabase.co")!
    private static let anonKey = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImhpeHltaWlkcHhubm92dG1zdmZwIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NzkyMjY1OTIsImV4cCI6MjA5NDgwMjU5Mn0.dkGd2pTtkz6euRVgMa6vXdOkHmV74M4nfXpY6Al3vbA"

    static func fetchLatest() async -> WidgetCard? {
        var components = URLComponents(
            url: supabaseURL.appendingPathComponent("rest/v1/cards"),
            resolvingAgainstBaseURL: false
        )!
        components.queryItems = [
            URLQueryItem(name: "select", value: "card_id,quote,works(title)"),
            URLQueryItem(name: "order", value: "card_id.desc"),
            URLQueryItem(name: "limit", value: "1"),
        ]

        var request = URLRequest(url: components.url!)
        request.setValue(anonKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(anonKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        struct Row: Decodable {
            let cardId: Int
            let quote: String
            let works: Work?
            struct Work: Decodable { let title: String }
        }

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
                return nil
            }
            let decoder = JSONDecoder()
            decoder.keyDecodingStrategy = .convertFromSnakeCase
            let rows = try decoder.decode([Row].self, from: data)
            guard let row = rows.first else { return nil }
            return WidgetCard(cardId: row.cardId, quote: row.quote, workTitle: row.works?.title ?? "")
        } catch {
            return nil
        }
    }
}

File: main_app/android/app/src/main/java/com/lifestyle/dailyscript/widget/WidgetDataLoader.kt
package com.lifestyle.dailyscript.widget

import com.lifestyle.dailyscript.BuildConfig
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

/**
 * iOS의 WidgetDataLoader.swift 와 동일한 역할.
 * 위젯은 앱의 Supabase 클라이언트 의존성을 가져오지 않기 위해
 * REST API 를 직접 HTTPURLConnection 으로 호출한다.
 */
internal data class WidgetCard(val quote: String, val workTitle: String)

internal object WidgetDataLoader {
    private val json = Json { ignoreUnknownKeys = true }

    @Serializable
    private data class Row(
        val quote: String,
        val works: Work? = null,
    ) {
        @Serializable
        data class Work(val title: String = "")
    }

    fun fetchLatest(): WidgetCard? {
        val base = BuildConfig.SUPABASE_URL.trimEnd('/')
        val anon = BuildConfig.SUPABASE_ANON_KEY
        if (base.isBlank() || anon.isBlank()) return null

        val select = URLEncoder.encode("quote,works(title)", "UTF-8")
        val urlStr = "$base/rest/v1/cards?select=$select&order=card_id.desc&limit=1"
        val conn = (URL(urlStr).openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            setRequestProperty("apikey", anon)
            setRequestProperty("Authorization", "Bearer $anon")
            setRequestProperty("Accept", "application/json")
            connectTimeout = 8000
            readTimeout = 8000
        }
        return try {
            if (conn.responseCode !in 200..299) return null
            val body = conn.inputStream.bufferedReader().use { it.readText() }
            val rows = json.decodeFromString<List<Row>>(body)
            val row = rows.firstOrNull() ?: return null
            WidgetCard(quote = row.quote, workTitle = row.works?.title.orEmpty())
        } catch (_: Throwable) {
            null
        } finally {
            conn.disconnect()
        }
    }
}
